### PR TITLE
Add Volcano gang scheduling support and adapter to AI conformance tests

### PR DESCRIPTION
**Pull request by krrish175-byte**
## Description
Following up on the initial gang scheduling tests implemented for Kueue, this PR adds support for testing [Volcano](https://volcano.sh/) as a gang scheduler in the AI Conformance test suite.

Due to Volcano's architecture, when a standard `batch/v1` Job is created, Volcano's Mutating Admission Webhook automatically generates a shadow `PodGroup` with `minMember: 1`, bypassing custom strict `minMember` requirements. This PR introduces an adapter to seamlessly override this behavior during the tests.

## Changes
- **Added `-gang-scheduler-name` flag**: Introduced a new CLI flag in `util_test.go` to specify the scheduler being tested (e.g. `volcano`).
- **Dynamic `PodGroup` Management**: Added an `applyVolcanoAdapter` helper in `gang_scheduling_test.go` that intercepts the generic `batch/v1` Job creation. When testing Volcano, it uses the dynamic client to programmatically create and clean up the `PodGroup` CRD alongside the job.
- **Webhook Override**: The adapter mutates the Job template to explicitly include the `scheduling.k8s.io/group-name` and `scheduling.volcano.sh/group-name` annotations pointing to our custom `PodGroup`. This overrides the Volcano mutating webhook, ensuring Volcano respects our strict `minMember: 1000` requirement for the negative tests.
- **E2E Script Update**: Updated `run-e2e-gcp.sh` to optionally install Volcano components when `GANG_SCHEDULER=volcano` is specified, removing the need for external policy engines or manual pre-provisioning.

## Testing
- Successfully verified both the `PositiveGangScheduling` and `NegativeGangScheduling` tests locally against a `kind` cluster with Volcano installed.
- Confirmed that the negative test pods successfully remain suspended/pending without partial scheduling when the `minMember` constraint cannot be met.

### CODE REVIEW

**Comment by krrish175-byte** [comment]
@ritazh PTAL

File: test/gang_scheduling_test.go [hunk 1]
```diff
@@ -254,3 +260,50 @@ func buildResourceList(cpuReq, memReq string) corev1.ResourceList {
 		corev1.ResourceMemory: resource.MustParse(memReq),
 	}
 }
+
```
Comment (janetkuo) [inline test/gang_scheduling_test.go]: Per the discussion in #80, `test/` should remain mechanism-agnostic and verify behavior. Embedding scheduler-specific CRD creation (`PodGroup`) directly in `gang_scheduling_test.go` couples the conformance suite to Volcano.

For Kueue, we pre-provisioned the `ResourceFlavor`, `ClusterQueue`, and `LocalQueue` in `test/e2e/run-e2e-gcp.sh` and configured the Job via flags. We should follow the same pattern here: pre-provision `PodGroup` objects (`pos-job` with minMember=2, `neg-job` with minMember=1000) in `test/e2e/run-e2e-gcp.sh` and configure the Job via generic flags (`-gang-scheduler-name`, pod annotations).

If dynamic adapters are needed in the future, they should implement a common interface registered in a separate adapter package rather than being inlined into the test flow.

File: test/gang_scheduling_test.go [hunk 2]
```diff
@@ -254,3 +260,50 @@ func buildResourceList(cpuReq, memReq string) corev1.ResourceList {
...
+	// 2. Create PodGroup
+	podGroup := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "scheduling.volcano.sh/v1beta1",
+			"kind":       "PodGroup",
+			"metadata": map[string]interface{}{
+				"name":      job.Name,
+				"namespace": job.Namespace,
+			},
+			"spec": map[string]interface{}{
```
Comment (janetkuo) [inline test/gang_scheduling_test.go]:
```suggestion
	var minMember int64 = 1
	if job.Spec.Parallelism != nil {
		minMember = int64(*job.Spec.Parallelism)
	}
	podGroup := &unstructured.Unstructured{
		Object: map[string]interface{}{
			"apiVersion": "scheduling.volcano.sh/v1beta1",
			"kind":       "PodGroup",
			"metadata": map[string]interface{}{
				"name":      job.Name,
				"namespace": job.Namespace,
			},
			"spec": map[string]interface{}{
				"minMember": minMember,
				"queue":     "default",
			},
		},
	}
```
Guard against nil `job.Spec.Parallelism`.

File: test/gang_scheduling_test.go [hunk 3]
```diff
@@ -254,3 +260,50 @@ func buildResourceList(cpuReq, memReq string) corev1.ResourceList {
...
+		Group:    "scheduling.volcano.sh",
+		Version:  "v1beta1",
+		Resource: "podgroups",
+	}
+
+	if _, err := dynamicClient.Resource(gvr).Namespace(job.Namespace).Create(ctx, podGroup, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Volcano PodGroup for job %s: %v", job.Name, err)
+	}
+
+	t.Cleanup(func() {
```
Comment (janetkuo) [inline test/gang_scheduling_test.go]: `t.Cleanup` executes LIFO. Because Job deletion was registered before `applyVolcanoAdapter`, this deletes the `PodGroup` before the `Job` deletion is initiated. Deleting the `PodGroup` while the Job/pods are still terminating can cause admission errors or orphan pods in Volcano.

File: test/util_test.go [hunk 4]
```diff
@@ -152,6 +156,23 @@ func deleteNamespaceAndWait(ctx context.Context, t *testing.T, c kubernetes.Inte
 	return nil
 }
 
+// getDynamicClient creates a dynamic Kubernetes client using the kubeconfig flag.
+func getDynamicClient(t *testing.T) dynamic.Interface {
```
Comment (janetkuo) [inline test/util_test.go]: Add `t.Helper()`. Also, factor out the duplicated `clientcmd` loading and `rest.Config` creation shared between `getClientset` and `getDynamicClient` into a single `getRESTConfig(t *testing.T) *rest.Config` helper.

File: test/e2e/run-e2e-gcp.sh [hunk 5]
```diff
@@ -331,10 +331,28 @@ EOF
 
   echo "Waiting for ClusterQueue to be active..."
   kubectl wait --for=condition=Active clusterqueue/e2e-cq --timeout=60s
+elif [ "${GANG_SCHEDULER}" = "volcano" ]; then
+  echo "Installing Volcano..."
```
Comment (janetkuo) [inline test/e2e/run-e2e-gcp.sh]:
```suggestion
  echo "Installing Volcano..."
  kubectl apply -f "https://raw.githubusercontent.com/volcano-sh/volcano/${VOLCANO_VERSION}/installer/volcano-development.yaml"
```
Define `VOLCANO_VERSION="${VOLCANO_VERSION:-v1.9.0}"` at the top of the script with `K8S_VERSION` and `GPU_OPERATOR_VERSION` instead of hardcoding `v1.9.0` inline.

File: test/e2e/run-e2e-gcp.sh [hunk 6]
```diff
@@ -331,10 +331,28 @@ EOF
...
+  kubectl rollout status deployment -n volcano-system volcano-controllers --timeout=5m
+  kubectl rollout status deployment -n volcano-system volcano-scheduler --timeout=5m
+
+  echo "Creating test namespace..."
+  kubectl create namespace ai-conformance-gang-scheduling --dry-run=client -o yaml | kubectl apply -f -
 fi
 
 REMOTE_STACK
 
+echo "Preparing go test arguments..."
```
Comment (janetkuo) [inline test/e2e/run-e2e-gcp.sh]:
```suggestion
if [ "${GANG_SCHEDULER}" = "kueue" ]; then
    GANG_LABELS="-gang-job-labels=kueue.x-k8s.io/queue-name=e2e-lq"
else
    GANG_LABELS=""
fi
```
The escaped quotes `\"` pass literal quote characters into the flag argument, which corrupts label parsing in `strings.Split`.

File: test/e2e/run-e2e-gcp.sh [hunk 7]
```diff
@@ -331,10 +331,28 @@ EOF
...
+if [ "${GANG_SCHEDULER}" = "kueue" ]; then
+    GANG_LABELS="-gang-job-labels=\"kueue.x-k8s.io/queue-name=e2e-lq\""
+else
```
Comment (janetkuo) [inline test/e2e/run-e2e-gcp.sh]:
```suggestion
if [ "${GANG_SCHEDULER}" = "kueue" ]; then
    GANG_LABELS="-gang-job-labels=kueue.x-k8s.io/queue-name=e2e-lq"
else
    GANG_LABELS=""
fi
```
The escaped quotes `\"` result in literal quote characters being passed to the flag on the remote VM, which breaks label parsing in `strings.Split`.

Also consider combining `GANG_ARGS="-gang-scheduler-name=${GANG_SCHEDULER} ${GANG_LABELS}"` so line 372 doesn't evaluate to a blank line with a trailing backslash when `GANG_LABELS` is empty.

File: test/gang_scheduling_test.go [hunk 8]
```diff
@@ -254,3 +260,50 @@ func buildResourceList(cpuReq, memReq string) corev1.ResourceList {
 		corev1.ResourceMemory: resource.MustParse(memReq),
 	}
```
Comment (janetkuo) [inline test/gang_scheduling_test.go]: Instead of calling `applyVolcanoAdapter` directly from tests and returning early when `*gangSchedulerName != "volcano"`, wrap adapter execution in a dispatcher:

```go
func applyGangSchedulerAdapter(ctx context.Context, t *testing.T, dynamicClient dynamic.Interface, job *batchv1.Job) {
	switch *gangSchedulerName {
	case "", "kueue":
		// Standard batch/v1 Job; no adapter needed.
	case "volcano":
		applyVolcanoAdapter(ctx, t, dynamicClient, job)
	default:
		t.Fatalf("Unsupported -gang-scheduler-name: %q", *gangSchedulerName)
	}
}
```

This validates `-gang-scheduler-name` (preventing silent fall-through on typos like `volcanoo`) and provides a clean hook for future schedulers (e.g., YuniKorn, Coscheduling).

File: test/util_test.go [hunk 9]
```diff
@@ -152,6 +156,23 @@ func deleteNamespaceAndWait(ctx context.Context, t *testing.T, c kubernetes.Inte
...
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
```
Comment (janetkuo) [inline test/util_test.go]: Add `t.Helper()`. Also, `getDynamicClient` and `getClientset` duplicate `clientcmd` loading and `rest.Config` building line-for-line. Suggest using a shared helper.

File: test/e2e/run-e2e-gcp.sh [hunk 10]
```diff
@@ -331,10 +331,28 @@ EOF
 
   echo "Waiting for ClusterQueue to be active..."
   kubectl wait --for=condition=Active clusterqueue/e2e-cq --timeout=60s
+elif [ "${GANG_SCHEDULER}" = "volcano" ]; then
+  echo "Installing Volcano..."
+  kubectl apply -f https://raw.githubusercontent.com/volcano-sh/volcano/v1.9.0/installer/volcano-development.yaml
```
Comment (janetkuo) [inline test/e2e/run-e2e-gcp.sh]:
```suggestion
  kubectl apply -f "https://raw.githubusercontent.com/volcano-sh/volcano/${VOLCANO_VERSION}/installer/volcano-development.yaml"
```
nit: Define `VOLCANO_VERSION="${VOLCANO_VERSION:-v1.9.0}"` at the top

File: test/e2e/run-e2e-gcp.sh [hunk 11]
```diff
@@ -331,10 +331,28 @@ EOF
...
+
+  echo "Creating test namespace..."
+  kubectl create namespace ai-conformance-gang-scheduling --dry-run=client -o yaml | kubectl apply -f -
 fi
 
 REMOTE_STACK
 
+echo "Preparing go test arguments..."
+if [ "${GANG_SCHEDULER}" = "kueue" ]; then
+    GANG_LABELS="-gang-job-labels=\"kueue.x-k8s.io/queue-name=e2e-lq\""
```
Comment (janetkuo) [inline test/e2e/run-e2e-gcp.sh]:
```suggestion
    GANG_LABELS="-gang-job-labels=kueue.x-k8s.io/queue-name=e2e-lq"
```

Also consider combining them as `GANG_ARGS="-gang-scheduler-name=${GANG_SCHEDULER} ${GANG_LABELS}"` so line 372 doesn't evaluate to a blank line with a trailing backslash when GANG_LABELS is empty.

File: test/gang_scheduling_test.go [hunk 12]
```diff
@@ -254,3 +260,50 @@ func buildResourceList(cpuReq, memReq string) corev1.ResourceList {
...
+	podGroup := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "scheduling.volcano.sh/v1beta1",
+			"kind":       "PodGroup",
+			"metadata": map[string]interface{}{
+				"name":      job.Name,
+				"namespace": job.Namespace,
+			},
+			"spec": map[string]interface{}{
+				"minMember": int64(*job.Spec.Parallelism),
```
Comment (janetkuo) [inline test/gang_scheduling_test.go]: need to guard against nil `job.Spec.Parallelism`

File: test/gang_scheduling_test.go [hunk 13]
```diff
@@ -254,3 +260,50 @@ func buildResourceList(cpuReq, memReq string) corev1.ResourceList {
...
+		Resource: "podgroups",
+	}
+
+	if _, err := dynamicClient.Resource(gvr).Namespace(job.Namespace).Create(ctx, podGroup, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Volcano PodGroup for job %s: %v", job.Name, err)
+	}
+
+	t.Cleanup(func() {
+		t.Logf("Cleaning up Volcano PodGroup %s...", job.Name)
+		_ = dynamicClient.Resource(gvr).Namespace(job.Namespace).Delete(ctx, job.Name, metav1.DeleteOptions{})
```
Comment (janetkuo) [inline test/gang_scheduling_test.go]: Avoid ignoring deletion errors; log if deletion fails with an error other than NotFound.

File: test/gang_scheduling_test.go [hunk 14]
```diff
@@ -67,6 +71,7 @@ func TestGangScheduling(t *testing.T) {
 		})
 
 		t.Logf("Creating positive test job %s...", jobName)
+		applyVolcanoAdapter(ctx, t, dynamicClient, job)
```
Comment (janetkuo) [inline test/gang_scheduling_test.go]: t.Cleanup executes in LIFO order, so PodGroup deletion will be executed before the Job deletion. We should call the adapter before registering the Job cleanup. 

Same for the NegativeGangScheduling test below

File: test/gang_scheduling_test.go [hunk 15]
```diff
@@ -254,3 +260,50 @@ func buildResourceList(cpuReq, memReq string) corev1.ResourceList {
 		corev1.ResourceMemory: resource.MustParse(memReq),
 	}
 }
+
+func applyVolcanoAdapter(ctx context.Context, t *testing.T, dynamicClient dynamic.Interface, job *batchv1.Job) {
+	if *gangSchedulerName != "volcano" {
+		return
```
Comment (janetkuo) [inline test/gang_scheduling_test.go]: We should validate `-gang-scheduler-name` to avoid a silent fall-through on unsupported scheduler name or typos. 

Instead of calling applyVolcanoAdapter directly, wrap this in a function that use switch-cases to validate `-gang-scheduler-name`, run adapter for volcano case, and default case to return an error on unsupported -gang-scheduler-name. This makes it easier to add future schedulers too.

File: test/util_test.go [hunk 16]
```diff
@@ -88,11 +91,12 @@ func init() {
 		"Comma-separated key=value labels to apply to the generic gang scheduling Job (e.g. kueue.x-k8s.io/queue-name=e2e-lq).")
 	gangNegativeWindow = flag.Duration("gang-negative-window", 30*time.Second,
 		"Duration to observe the negative gang scheduling test job to verify no pods are partially scheduled.")
+	gangSchedulerName = flag.String("gang-scheduler-name", "",
+		"Name of the gang scheduler being tested (e.g. 'volcano'). Used to apply adapter logic if required.")
```
Comment (puja108) [inline test/util_test.go]:
```suggestion
		"Name of the gang scheduler being tested (currently supports: 'kueue' or 'volcano'). Used to apply adapter logic if required.")
```

making it more concrete here and staying agnostic to not mention just one of them, IMHO just listing both is currently fine, but we can change that if we have too many solutions to list

Comment (krrish175-byte) [reply test/util_test.go]: PTAL

File: test/gang_scheduling_test.go [hunk 17]
```diff
@@ -254,3 +266,67 @@ func buildResourceList(cpuReq, memReq string) corev1.ResourceList {
...
+
+func applyGangSchedulerAdapter(ctx context.Context, t *testing.T, dynamicClient dynamic.Interface, job *batchv1.Job) error {
+	switch *gangSchedulerName {
+	case "volcano":
+		applyVolcanoAdapter(ctx, t, dynamicClient, job)
+		return nil
+	case "kueue":
+		// Kueue handles gang scheduling automatically via annotations/labels, no extra API resources needed here.
+		return nil
+	default:
```
Comment (janetkuo) [inline test/gang_scheduling_test.go:278]: This fails when `-gang-scheduler-name` is omitted (default to `""`). Conformance tests must be runnable using standard defaults without requiring every optional flag to be explicitly supplied. Let's set the default flag value to `"kueue"`.

File: test/gang_scheduling_test.go [hunk 18]
```diff
@@ -254,3 +266,67 @@ func buildResourceList(cpuReq, memReq string) corev1.ResourceList {
...
+	}
+
+	gvr := schema.GroupVersionResource{
+		Group:    "scheduling.volcano.sh",
+		Version:  "v1beta1",
+		Resource: "podgroups",
+	}
+
+	if _, err := dynamicClient.Resource(gvr).Namespace(job.Namespace).Create(ctx, podGroup, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Volcano PodGroup for job %s: %v", job.Name, err)
```
Comment (janetkuo) [inline test/gang_scheduling_test.go:322]: Return error from applyVolcanoAdapter back through applyGangSchedulerAdapter, and add t.Helper() to both helpers so call-site locations are preserved in test failure output.

File: test/e2e/run-e2e-gcp.sh
Comment (janetkuo) [inline test/e2e/run-e2e-gcp.sh:259]: nit: given that we support volcano now, change it to generic "Gang Scheduler"

File: test/gang_scheduling_test.go [hunk 19]
```diff
@@ -254,3 +266,67 @@ func buildResourceList(cpuReq, memReq string) corev1.ResourceList {
 		corev1.ResourceMemory: resource.MustParse(memReq),
 	}
 }
+
+func applyGangSchedulerAdapter(ctx context.Context, t *testing.T, dynamicClient dynamic.Interface, job *batchv1.Job) error {
```
Comment (janetkuo) [inline test/gang_scheduling_test.go:270]: Add unit tests verifying adapter dispatch for empty string, "kueue", "volcano", and invalid inputs

**Review by janetkuo — COMMENTED** [review]
Mostly looks good. A few final comments

File: test/gang_scheduling_unit_test.go [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+				if got := err.Error(); !containsSubstring(got, tt.errSubstring) {
+					t.Errorf("error %q does not contain %q", got, tt.errSubstring)
+				}
+			}
+		})
+	}
+}
+
+// Note: The "volcano" case requires a dynamic client and is tested via
+// integration tests rather than unit tests.
```
Comment (janetkuo) [inline test/gang_scheduling_unit_test.go]: [Non-blocking]: You can test the "volcano" case without an integration cluster by passing `dynamicfake.NewSimpleDynamicClient(runtime.NewScheme())` to `applyGangSchedulerAdapter`. This allows verifying that `job.Spec.Template.Spec.SchedulerName`, `group-name` annotations, and the `PodGroup` unstructured resource are configured as expected.

File: test/gang_scheduling_unit_test.go [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+
+func containsSubstring(s, sub string) bool {
+	return len(s) >= len(sub) && (s == sub || len(s) > 0 && containsIndex(s, sub))
```
Comment (janetkuo) [inline test/gang_scheduling_unit_test.go:74]: Just use standard `strings.Contains(s, sub)` instead of custom `containsSubstring` and `containsIndex` helpers

**Review by janetkuo — APPROVED** [review]
/lgtm

Just a nit

/hold
for others to take a look as well

File: test/e2e/run-e2e-gcp.sh [hunk 22]
```diff
@@ -31,6 +31,7 @@ GCE_IMAGE_FAMILY="${GCE_IMAGE_FAMILY:-common-cu129-ubuntu-2404-nvidia-580}"
 GCE_IMAGE_PROJECT="${GCE_IMAGE_PROJECT:-deeplearning-platform-release}"
 K8S_VERSION="${K8S_VERSION:-v1.35.0}"
 GPU_OPERATOR_VERSION="${GPU_OPERATOR_VERSION:-v26.3.1}"
+VOLCANO_VERSION="${VOLCANO_VERSION:-v1.9.0}"
```
Comment (ritazh) [inline test/e2e/run-e2e-gcp.sh:34]: [The linked compatibility table first ](https://github.com/volcano-sh/volcano/blob/master/README.md#kubernetes-compatibility) lists Kubernetes 1.35 support in Volcano 1.15. Use a compatible 1.15.x release or lower the Kubernetes version.

File: test/e2e/run-e2e-gcp.sh [hunk 23]
```diff
@@ -331,10 +332,29 @@ EOF
 
   echo "Waiting for ClusterQueue to be active..."
   kubectl wait --for=condition=Active clusterqueue/e2e-cq --timeout=60s
+elif [ "${GANG_SCHEDULER}" = "volcano" ]; then
```
Comment (ritazh) [inline test/e2e/run-e2e-gcp.sh:335]: The current Prow job invokes run-e2e-gcp.sh without GANG_SCHEDULER, so the script defaults to Kueue and never enters the Volcano installation branch. The new unit test also explicitly omits the Volcano case. which means neither the PodGroup creation nor actual Volcano scheduling is covered by the successful check. Is the plan to add a Volcano Prow variant and a fake-dynamic-client unit case that verifies scheduler name, annotations, GVR, and spec.minMember?

**Comment by krrish175-byte** [comment]
@ritazh PTAL

File: test/e2e/run-e2e-gcp.sh (hunk 23 above)
Comment (ritazh) [reply test/e2e/run-e2e-gcp.sh]: Thanks for adding `TestApplyVolcanoAdapter` to cover unit-test. The current presubmit and periodic Prow configurations still do not set GANG_SCHEDULER=volcano, so passing E2E check only exercises Kueue. If we want CI to also include volcano, then need to add a test-infra PR to add a Volcano variant.
